sdks/python/apache_beam/yaml/yaml_mapping.py

```diff
@@ -27,33 +28,140 @@
 from apache_beam.yaml import yaml_provider
 
 
-def _as_callable(original_fields, expr):
-  if expr in original_fields:
-    return expr
+# TODO(yaml) Consider adding optional language version parameter to support ECMAScript 5 and 6
+def expand_mapping_func(
+    transform_name,
+    language,
+    original_fields,
+    expression=None,
+    callable=None,
+    path=None,
+    name=None):
+
+  # Argument checking
+  if not expression and not callable and not path and not name:
+    raise ValueError(
+        f'{transform_name} must specify either "expression", "callable", or both "path" and "name"'
+    )
+  if expression and callable:
+    raise ValueError(
+        f'{transform_name} cannot specify "expression" and "callable"')
+  if (expression or callable) and (path or name):
+    raise ValueError(
+        f'{transform_name} cannot specify "expression" or "callable" with "path" or "name"'
+    )
+  if path and not name:
+    raise ValueError(f'{transform_name} cannot specify "path" without "name"')
+  if name and not path:
+    raise ValueError(f'{transform_name} cannot specify "name" without "path"')
+
+  def _get_udf_from_file():
+    # Local UDF file case
+    if not path.startswith("gs://"):
+      try:
+        with open(path, 'r') as local_udf_file:
+          return local_udf_file.read()
+      except Exception as e:
+        raise IOError(f'Error opening file "{path}": {e}')
+
+    # GCS UDF file case
+    from google.cloud import storage
+
+    # Parse GCS file location
+    gcs_file_parts = str(path[5:]).split('/')
+    gcs_bucket_name = gcs_file_parts[0]
+    gcs_folder = '/'.join(gcs_file_parts[1:])
+
+    # Instantiates a client and downloads file to string
+    storage_client = storage.Client()
+    bucket = storage_client.get_bucket(gcs_bucket_name)
+    blob = bucket.blob(gcs_folder)
+    gcs_file = blob.download_as_string().decode('utf-8')
+
+    return gcs_file
+
+  def _check_file_ext(ext):
+    if not path.endswith(ext):
+      raise ValueError(f'File "{path}" is not a valid {ext} file.')
+
+  def _js2py_import():
+    return (['  try:'] + ['    import js2py'] + ['  except ImportError:'] + [
+        '    raise ImportError("js2py must be installed to run javascript UDF\'s for YAML mapping transforms.")'
+    ])
+
+  # Javascript UDF case
+  if language == 'javascript':
+    # Javascript expression case
+    if expression:
+      parameters = [name for name in original_fields if name in expression]
+      js_func = 'function fn(' + ','.join(
+          [name for name in parameters]) + '){' + 'return (' + expression + ')}'
+      source = '\n'.join(['def fn(__row__):'] + _js2py_import() +
```

Comment:
We should lift the import out of the function definition. (We should probably look at lifting the translation to happen once rather than on ever call as well. This could be done with

```args_from_row = ', '.join(f'__row__.{name}' for name in parameters)```

then generating the source

```
js_callable = js2py.eval_js({repr(js_func)})
def fn(__row__):
  return js_callable({args_from_row})
```